### ISSUE TITLE
Oddly timed GC

### ISSUE THREAD

**Issue by Hixie** [flutter/flutter#74668]
A customer filed this bug with a trace showing an unexpectedly-timed GC: http://b/178207429

We should examine the trace to see if there's anything we can improve based on it.

**Comment by zanderso** [comment]
@dnfield should this be grouped with the similar issue about the priority of Dart worker threads?

**Comment by zanderso** [comment]
In particular https://github.com/flutter/flutter/issues/70119 for https://buganizer.corp.google.com/issues/171917546

**Comment by dnfield** [comment]
I think this is the same as the worker thread priority issue.

The thing is, the GC is running concurrently there. It was started during some idle time, but the concurrent workers continued to mark away for hundreds of ms. They're doing heap compaction work in parallel.

I filed https://github.com/flutter/flutter/issues/72387 and https://github.com/flutter/flutter/issues/70636 to try to look at issues related to this.

My preliminary findings are that moving to one worker thread is bad for performance - Dart is able to do some additional optimizations when it uses two threads, and I consistently got slightly worse results (on a more high powered phone than our benchmarks) when using only one thread.

We've had other reports from customers that slightly lowering the priority of the worker threads compared to the UI/Raster thread is beneficial here.

**Comment by Hixie** [comment]
Lowering the priority of those threads at least while we're getting a frame ready, assuming we're not in a dire "need to GC now or we're immediately doomed" state, seems like an excellent idea.

**Comment by dnfield** [comment]
For that we really need something like https://github.com/dart-lang/sdk/issues/44228

**Comment by aam** [comment]
What is the worry here? The concurrent marking is running for ~100ms on two worker threads, it doesn't block anything else, should not cause any user-visible jank or slowdown.

**Comment by Hixie** [comment]
I may be misreading the trace but it looks like it's causing the frame rendering to be descheduled a bit, thus making the frame take longer than usual.

**Comment by tobinharris** [comment]
Apologies if this is the wrong place to chip in, but I'm struggling with a freeze for 5-30 seconds in my app ([bug report](https://github.com/flutter/flutter/issues/82641)). Every time, I can see the profiler shows a GC and parallel mark running in the UI, and another parallel mark running on a separate Dart Worker. Could there be any kind of race condition? 

I don't have the skills to dig deeper into what's happening, and for some reason my CPU profiling is always empty even though I'm doing `--profile` and connecting to an iOS device.

**Comment by Hixie** [comment]
@tobinharris just saw your comment -- please file new issues when dealing with new problems even if they look very similar to existing problems, it's better to start with separate issues then mark them duplicates of each other than to try to manage multiple problems in one issue.

**Comment by Hixie** [comment]
I'm going to close this in favour of https://github.com/flutter/flutter/issues/89884 which is more actionable.
